# Now You Code In Class: Tricks of The Pandas Masters

We will try something a bit different in the Activity - A series of Pandas coding challenges!


For this Now You Code, you will need the Zomato api https://developers.zomato.com/api which provides API access to local area restaurants information.


In [1]:
import pandas as pd
import import_ipynb
import googleapi as google
import requests

pd.set_option('display.max_colwidth', None)
google_api_key = ''

importing Jupyter notebook from googleapi.ipynb


## Warm-Up 1

Load the following data set into a data frame, called `exam_df`

https://raw.githubusercontent.com/mafudge/datasets/master/exam-scores/exam-scores.csv

HINT: observe the file format!

In [ ]:
# one liner
file = "https://raw.githubusercontent.com/mafudge/datasets/master/exam-scores/exam-scores.csv"
exam_df = ????

In [ ]:
from IPython.display import display
display(exam_df)

## Warm-Up 2

Load the following HTML table into a data frame called `res_df`

https://en.wikipedia.org/wiki/Graphics_display_resolution#High-definition

or try this one getting diamondbacks hitting

https://www.espn.com/mlb/boxscore/_/gameId/401430916


HINT: Which table is it on the page?

NameError: name 'pd' is not defined

## Warm-Up 3

For your `exam_df` display exams where the students made their own study guide  and finished in 30 minutes or less.

HINT: This can be multiple steps


## Warm-Up 4

For your `res_df`, display the displays which are larger than 1920 x 1080. Show columns `Name` and `H:V` only

HINT: Again, think multiple steps

## Challenge 1

1. Search the Open Street Maps  geocode api for "Syracuse" (leave off the NY) put the results in a dataframe (there should be several results). 

Example:
https://nominatim.openstreetmap.org/search?q=Syracuse&format=json


In [14]:
import requests
from coursetools.display import display_dataframes
url = "https://nominatim.openstreetmap.org/search"
params = { "q" : "Syracuse", "format" : "json" }
headers = { 'User-Agent' : 'IST256 app'}
response = requests.get(url, params = params, headers=headers)
response.raise_for_status()
results = response.json()
df1 = pd.DataFrame(results) # construct dataframe from the list of dict
df2 = pd.read_json(response.url) # read the url directly ()
display_dataframes(df1,df2)

In [10]:
df = pd.DataFrame(results)
df

place_id  \
0  298364595   
1  297907245   
2  297997314   
3  298019806   
4  298049216   
5   67220594   
6  298076223   
7  298559959   
8  298580155   
9  297914225   

                                                                  licence  \
0  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
1  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
2  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
3  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
4  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
5  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
6  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
7  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
8  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
9  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   

   osm_type      osm_id                                         boundingbox  \
0  relation      174916      [42.984165, 43.086121, -76.204629, -76.074273]   
1  relation       39169    [36.9477054, 37.1156703, 15.0829726, 15.3367251]   
2  relation      129906    [37.941272, 37.990492, -101.816776, -101.736708]   
3  relation      169314     [40.651206, 40.6826315, -96.197821, -96.172004]   
4  relation      198737   [41.0529648, 41.11093, -112.1159667, -112.035662]   
5      node  6241108067  [43.0718312, 43.0818312, -76.1744003, -76.1644003]   
6  relation      141314       [38.665793, 38.67337, -92.883419, -92.869404]   
7  relation      127610       [41.403028, 41.43607, -85.768612, -85.734116]   
8  relation      182667    [38.9922728, 39.007177, -81.9841277, -81.943146]   
9  relation       39155    [36.6441671, 37.4128826, 14.7729261, 15.3367251]   

                  lat                 lon  \
0          43.0481221         -76.1474244   
1          37.0646139          15.2907196   
2          37.9805686         -101.752671   
3           40.657224          -96.186398   
4          41.0895938        -112.0638903   
5          43.0768312         -76.1694003   
6          38.6700209         -92.8732509   
7          41.4278409         -85.7524957   
8          38.9975789         -81.9740274   
9  37.029827850000004  15.086786626415927   

                                                                                         display_name  \
0                                          City of Syracuse, Onondaga County, New York, United States   
1                                                                           Siracusa, Sicilia, Italia   
2                                             Syracuse, Hamilton County, Kansas, 67878, United States   
3                                                      Syracuse, Otoe County, Nebraska, United States   
4                                                  Syracuse, Davis County, Utah, 84075, United States   
5  Syracuse, 131, NBT Bank Parkway, City of Syracuse, Onondaga County, New York, 13208, United States   
6                                                    Syracuse, Morgan County, Missouri, United States   
7                                                  Syracuse, Kosciusko County, Indiana, United States   
8                                                  Syracuse, Meigs County, Ohio, 45779, United States   
9                                                                           Siracusa, Sicilia, Italia   

      class            type  importance  \
0  boundary  administrative    0.676776   
1  boundary  administrative    0.598054   
2  boundary  administrative    0.535495   
3  boundary  administrative    0.461126   
4  boundary  administrative    0.436855   
5   railway         station    0.427252   
6  boundary  administrative    0.420439   
7  boundary  administrative    0.420439   
8  boundary  administrative    0.409042   
9  boundary  administrative

In [9]:
df = pd.read_json(response.url)
df

place_id  \
0  298364595   
1  297907245   
2  297997314   
3  298019806   
4  298049216   
5   67220594   
6  298076223   
7  298559959   
8  298580155   
9  297914225   

                                                                  licence  \
0  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
1  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
2  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
3  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
4  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
5  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
6  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
7  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
8  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   
9  Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright   

   osm_type      osm_id                                         boundingbox  \
0  relation      174916      [42.984165, 43.086121, -76.204629, -76.074273]   
1  relation       39169    [36.9477054, 37.1156703, 15.0829726, 15.3367251]   
2  relation      129906    [37.941272, 37.990492, -101.816776, -101.736708]   
3  relation      169314     [40.651206, 40.6826315, -96.197821, -96.172004]   
4  relation      198737   [41.0529648, 41.11093, -112.1159667, -112.035662]   
5      node  6241108067  [43.0718312, 43.0818312, -76.1744003, -76.1644003]   
6  relation      141314       [38.665793, 38.67337, -92.883419, -92.869404]   
7  relation      127610       [41.403028, 41.43607, -85.768612, -85.734116]   
8  relation      182667    [38.9922728, 39.007177, -81.9841277, -81.943146]   
9  relation       39155    [36.6441671, 37.4128826, 14.7729261, 15.3367251]   

         lat         lon  \
0  43.048122  -76.147424   
1  37.064614   15.290720   
2  37.980569 -101.752671   
3  40.657224  -96.186398   
4  41.089594 -112.063890   
5  43.076831  -76.169400   
6  38.670021  -92.873251   
7  41.427841  -85.752496   
8  38.997579  -81.974027   
9  37.029828   15.086787   

                                                                                         display_name  \
0                                          City of Syracuse, Onondaga County, New York, United States   
1                                                                           Siracusa, Sicilia, Italia   
2                                             Syracuse, Hamilton County, Kansas, 67878, United States   
3                                                      Syracuse, Otoe County, Nebraska, United States   
4                                                  Syracuse, Davis County, Utah, 84075, United States   
5  Syracuse, 131, NBT Bank Parkway, City of Syracuse, Onondaga County, New York, 13208, United States   
6                                                    Syracuse, Morgan County, Missouri, United States   
7                                                  Syracuse, Kosciusko County, Indiana, United States   
8                                                  Syracuse, Meigs County, Ohio, 45779, United States   
9                                                                           Siracusa, Sicilia, Italia   

      class            type  importance  \
0  boundary  administrative    0.676776   
1  boundary  administrative    0.598054   
2  boundary  administrative    0.535495   
3  boundary  administrative    0.461126   
4  boundary  administrative    0.436855   
5   railway         station    0.427252   
6  boundary  administrative    0.420439   
7  boundary  administrative    0.420439   
8  boundary  administrative    0.409042   
9  boundary  administrative    0.404285   

                                                                                   icon  
0  https://nominatim.openstreetmap.org/ui/mapicons/poi_boundary_administrative.p

## Challenge 2

1. Load the locations into a dataframe. using the `pd.DataFrame()` method. Start with whatever is of type `list` as that will automatically place it into a `DataFrame` 

## Challenge 3

1. Let's call the IST256 google places search api, looking for bars in the Syracuse NY area, and load the results into a dataframe. You can call `google.googlePlacesSearch()` With these inputs lat=`43.0481`, lng=`-76.1474`, type=`'bar'`



## Challenge 4

Notice the nested JSON. Yuk. This can be fixed using `pd.json_normalize` instead of `pd.DataFrame`.

Fix the results above so that the nested JSON is flattened in the dataframe.


## Challenge 5

Convert your dataframe back into a list of dict.

1. use `to_records()` to convert the dataframe in Challenge 4 back into a `list` of `dict`.
2. loop over the rows and print the bar name, rating, and whether it is open now, 
   for only those rows where the bar is open now.

NOTE: The keys match he columns in the data frame.



In [ ]:
# run this code to turn in your work!
from coursetools.submission import Submission
Submission().submit_now()